### ISSUE TITLE
LM AOT Spiral Inaccuracy

### ISSUE THREAD

**Issue by ggalfi** [dseagrav/NASSP#347]
Hi folks,

I was experimenting with the cursor-spiral technique during P52, and I found it's accuracy is even worse than coarse alignment methods (in some cases I got for V06N05 a value more than +005.00, which is definitely unacceptable). I found out that my cursor measurements are quite good but the spiral ones are usually off up to a few degrees. I investigated the problem and discovered two sources of the inaccuracy:
1. This applies only on my machine (or on my fairly old monitor), because I cannot use orbitersim with more than 1024p vertical resolution. The problem is, that the AOT reticle needs 1050 pixels, so it will not fit into my 1024p screen. However the preset  60 degrees FOV is the screen's FOV and not the reticle's as it should be! So on my monitor the AOT FOV appears with a little bit more than 60 degrees which certainly explains some of the error in my spiral measurements. I changed the line in lemvc.cpp from
oapiCameraSetAperture(RAD * 30.0);
which simply set screen FOV to 60 to
DWORD w, h;
oapiGetViewportSize(&w, &h);
oapiCameraSetAperture(atan(tan(RAD * 30.0) * h/1050.0));
which will adapt the screen aperture in a way that it will align 1050 pixels to 60 degrees of FOV.
2. Although the change described above improved the accuracy (V06N05 was scattering around +000.70) still i wasn't satisfied with it. I understood that in calculations the LGC using the spiral's distance from the center as an angular value. For example if you rotate the reticle with 270 degrees after cursor measurement of a star to align the spiral on it, that means you see the star 3/4 way of the center or - more preciselly - the angle between the viewing direction and the star is 0.75 * 30=22.5 degrees. However the Orbitersim - as most of the 3d games - uses some sort of planar projection so there is no simple linear correspondence between pixel distance and angular distance as it is assumed in the current code for drawing the spiral. I derived the exact the formula for that (some trigonometry helps here) and I changed the way of calculation of the spiral's radial coordinate in lempanel.cpp from
b = - RETICLE_RADIUS / (2*PI);
r = b * theta;
to the corrected version
b = -RETICLE_RADIUS / tan(2 * PI *30/ 360);
r = b * tan(theta * 30/360);
That is the reason behind that I used that atan-tan formula in oapiCameraSetAperture instead of simply calculating 30 * h/1050 for desired FOV.
There is a significant difference between the original (red) and corrected (green) spiral as you can see on screenshot below:
![image](https://user-images.githubusercontent.com/33571169/39252577-419c5a26-48a6-11e8-9bed-1c20aae8d633.png)

With these changes I was able to go down to +000.03 in star angle difference which is still not the "all balls" result I can sometimes achieve with the sextant and the CMC but I think I will not miss the Moon with this alignment if I attempt a landing :)

I attached the codes which contains my adjustments to the 22nd April version.
[aot_spiral_corrected.zip](https://github.com/dseagrav/NASSP/files/1947812/aot_spiral_corrected.zip)

**Comment by indy91** [comment]
Oh wow, this is something we have been issues with for quite a while. We just couldn't figure out what was wrong with the spirale code or the AOT view. Your solution sounds very plausible, thanks so much for working that all out! I'll get to work to test and implement this.

**Comment by jalexb88** [comment]
Great find! Also does your monitor give any display issues with the LPD panel? It is 1080p vertical.

**Comment by indy91** [comment]
Initial test resulted in +000.06 star angle difference. I'm pretty rusty when it comes to P57s and that is much better than what I usually got with option 2 (two celestial bodies) before. The +000.70 you mentioned was about the error we were all getting before with P57. So I'm quite confident that you figured out this issue, which we've been having for a quite a while. Thanks again! 

Feel free to drop by anytime in the NASSP IRC channel, #nassp on freenode. That's where NASSP developers and users are hanging out most of the time.

**Comment by indy91** [comment]
So one issue with the variable FOV are resolutions beyond 1050 pixels in height, like 1920x1080. The bitmap actually gets stretched in that case, so an maybe an upper limit for the aperture calculation has to be used in that case?

**Comment by ggalfi** [comment]
@jalexb88 Previously I've only thrown a few (semi)manual landing in P67 mode after pitchover, so I didn't care too much of the LPD. But as you asked, I played a bit with the Apollo 17 PDI scenario in the WIP section and let the PNGS do the job till touchdown. And yes, if I set the screen FOV to 57.39, to align the 1080p with the 60 degrees FOV then the LPD value on the DSKY is within a +-1 degree match with the future landing point seen on the LPD scale. I am quite surprised if not amazed on the accuracy of this whole stuff :) 
Anyway I'd greatly appreciate if a similar adaptive FOV calculation for the LPD as I have made for the AOT reticle would be included in an upcoming release.

**Comment by ggalfi** [comment]
@indy91: yes, I considered that issue with stretching but I wasn't sure how Orbitersim handles the smaller-than-resolution bitmaps. But if it handles in the way you described then instead of
oapiCameraSetAperture(atan(tan(RAD * 30.0) * h/1050.0));
this could be used
oapiCameraSetAperture(atan(tan(RAD * 30.0)  * min(h/1050.0, 1.0) ));
and everybody will be happy :)

**Comment by jalexb88** [comment]
The code seems to work, even at very high resolution such as my UHD monitor (3840x2160) the AOT window is stretched to fill the entire height of the screen and FOV stays at 60. I also tried the new FOV handling on the LPD panel, but with oapiCameraSetAperture(atan(tan(RAD * 30.0) * min(h/**1050**.0, 1.0) )); as that bitmap is 1080p, then tried with lower then 1080p resolutions to see if FOV is lowered correctly and sure enough it works nicely.

**Comment by ggalfi** [comment]
Tested the a584212 commit and the AOT measurements are fine on my machine.

I've seen that the FOV calculation for the LPD window is not changed yet, and I suggest to leave as it is (contrary to my previous request). The reason is that I discovered in the LM Timeline checklist that the P52 has an LPD calibration feature, so the LGC should cure a few percent FOV deviance by itself.

I am happy to close this issue now. Thanks for your efforts put into this nice and enermous project!
